lsp-mode.el

```diff
@@ -3676,16 +3681,42 @@ The method uses `replace-buffer-contents'."
                                     beg (+ beg (length new-text))
                                     length)))))))))
 
-(defun lsp--indent-snippets? ()
-  "Enable indenting of snippets for everything but `org-mode'.
+(defun lsp--to-yasnippet-snippet (snippet)
+  "Convert LSP SNIPPET to yasnippet snippet."
+  ;; LSP snippet doesn't escape "{", but yasnippet requires escaping it.
+  (replace-regexp-in-string (rx (or bos (not (any "$" "\\"))) (group "{"))
+                            (rx "\\" (backref 1))
+                            snippet
+                            nil nil 1))
 
-Indending snippets is extremely slow in `org-mode' buffers since
-it has to calculate indentation based on SRC block position."
-  (unless (derived-mode-p 'org-mode)
-    'auto))
+(defvar-local lsp-enable-relative-indentation nil
+  "Enable relative indentation when insert texts, snippets ... from language server.")
+
+(defun lsp--expand-snippet (snippet &optional start end expand-env keep-whitespace)
+  "Wrapper of `yas-expand-snippet' with all of it arguments.
+The snippet will be convert to LSP style and indent according to
+LSP server result."
+  (let* ((inhibit-field-text-motion t)
+         (offset (save-excursion
+                   (goto-char start)
+                   (back-to-indentation)
+                   (buffer-substring-no-properties
+                    (line-beginning-position)
+                    (point))))
+         (yas-indent-line (unless keep-whitespace 'auto))
+         (yas-also-auto-indent-first-line nil)
+         (indent-line-function (if (or lsp-enable-relative-indentation
```

Comment:
shouldn't we have keep-whitespace here as well?

Reply:
No need, the `yas-indent-line` will decide to use `indent-line-function` or not.
In case of `keep-whitespace`, it will insert as it without using `indent-line-function`

Reply:
ok, I am not sure I understand what you are saying - we will test it more extensively when servers start to support that feature, thanks.

Reply:
So the gist of it is that `yas-indent-line` when set to `auto` will indent each line of snippet with `indent-according-to-mode`, which in turn will call `indent-line-function`.
In case `yas-indent-line` is set to `nil`, it will not apply any snippet-side indentation.
If you notice a few lines upper, it's set as `(yas-indent-line (unless keep-whitespace 'auto))`

Reply:
I see - I though that keep-whitespace = apply relative-indentation.